<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="date" content="2018-06-27" />

<title>NCBI Database Retrieval</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">NCBI Database Retrieval</h1>
<h4 class="date"><em>2018-06-27</em></h4>



<div id="retrieve-sequence-databases-from-ncbi" class="section level1">
<h1>Retrieve Sequence Databases from NCBI</h1>
<p>NCBI stores a variety of specialized database such as <a href="https://www.ncbi.nlm.nih.gov/guide/data-software/">Genbank, RefSeq, Taxonomy, SNP, etc.</a> on their servers. The <code>download.database()</code> and <code>download.database.all()</code> functions implemented in <code>biomartr</code> allows users to download these databases from NCBI. This process might be very useful for downstream analyses such as sequence searches with e.g. BLAST. For this purpose see the R package <a href="https://github.com/HajkD/metablastr">metablastr</a> which aims to seamlessly inegrate <code>biomartr</code> based genomic data retrieval with downsteam large-scale BLAST searches.</p>
<ul>
<li><a href="#ist-available-databases">1. List available NCBI databases with <code>listNCBIDatabases()</code></a></li>
<li><a href="#download-ncbi-databases">2. Download NCBI databases with <code>download.database.all()</code></a>
<ul>
<li><a href="#example-ncbi-nr">2.1 Example NCBI nr retrieval</a></li>
<li><a href="#example-ncbi-nt">2.2 Example NCBI nt retrieval</a></li>
<li><a href="#example-ncbi-refseq">2.3 Example NCBI RefSeq retrieval</a></li>
<li><a href="#example-pdb">2.4 Example Protein Database (PDB) retrieval</a></li>
<li><a href="#example-ncbi-taxonomy">2.5 Example NCBI Taxonomy retrieval</a></li>
<li><a href="#example-ncbi-swissprot">2.6 Example NCBI Swissprot retrieval</a></li>
<li><a href="#example-ncbi-cdd-delta">2.7 Example NCBI CDD Delta retrieval</a></li>
</ul></li>
</ul>
<div id="list-available-databases" class="section level2">
<h2>List available databases</h2>
<p>Before downloading specific databases from NCBI users might want to list available databases. Using the <code>listNCBIDatabases()</code> function users can retrieve a list of available databases stored on NCBI.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># retrieve a list of available sequence databases at NCBI</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;all&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;16SMicrobial&quot;            &quot;cdd_delta&quot;               &quot;cloud&quot;                  
 [4] &quot;env_nr&quot;                  &quot;env_nt&quot;                  &quot;est&quot;                    
 [7] &quot;est_human&quot;               &quot;est_mouse&quot;               &quot;est_others&quot;             
[10] &quot;FASTA&quot;                   &quot;gene_info&quot;               &quot;gss&quot;                    
[13] &quot;gss_annot&quot;               &quot;htgs&quot;                    &quot;human_genomic&quot;          
[16] &quot;landmark&quot;                &quot;nr&quot;                      &quot;nt&quot;                     
[19] &quot;other_genomic&quot;           &quot;pataa&quot;                   &quot;patnt&quot;                  
[22] &quot;pdbaa&quot;                   &quot;pdbnt&quot;                   &quot;ref_prok_rep_genomes&quot;   
[25] &quot;ref_viroids_rep_genomes&quot; &quot;ref_viruses_rep_genomes&quot; &quot;refseq_genomic&quot;         
[28] &quot;refseq_protein&quot;          &quot;refseq_rna&quot;              &quot;refseqgene&quot;             
[31] &quot;sts&quot;                     &quot;swissprot&quot;               &quot;taxdb&quot;                  
[34] &quot;tsa_nr&quot;                  &quot;tsa_nt&quot;                  &quot;vector&quot;</code></pre>
<p>However, in case users already know which database they would like to retrieve they can filter for the exact files by specifying the NCBI database name. In the following example all sequence files that are part of the <code>NCBI nr</code> database shall be retrieved.</p>
<p>First, the <code>listNCBIDatabases(db = &quot;nr&quot;)</code> allows to list all files corresponding to the <code>nr</code> database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all NCBI nr files</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;nr&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;nr.00.tar.gz&quot; &quot;nr.83.tar.gz&quot; &quot;nr.75.tar.gz&quot; &quot;nr.01.tar.gz&quot;
 [5] &quot;nr.02.tar.gz&quot; &quot;nr.03.tar.gz&quot; &quot;nr.04.tar.gz&quot; &quot;nr.05.tar.gz&quot;
 [9] &quot;nr.76.tar.gz&quot; &quot;nr.16.tar.gz&quot; &quot;nr.54.tar.gz&quot; &quot;nr.55.tar.gz&quot;
[13] &quot;nr.56.tar.gz&quot; &quot;nr.06.tar.gz&quot; &quot;nr.15.tar.gz&quot; &quot;nr.30.tar.gz&quot;
[17] &quot;nr.57.tar.gz&quot; &quot;nr.58.tar.gz&quot; &quot;nr.59.tar.gz&quot; &quot;nr.60.tar.gz&quot;
[21] &quot;nr.61.tar.gz&quot; &quot;nr.62.tar.gz&quot; &quot;nr.63.tar.gz&quot; &quot;nr.64.tar.gz&quot;
[25] &quot;nr.65.tar.gz&quot; &quot;nr.66.tar.gz&quot; &quot;nr.67.tar.gz&quot; &quot;nr.68.tar.gz&quot;
[29] &quot;nr.69.tar.gz&quot; &quot;nr.70.tar.gz&quot; &quot;nr.71.tar.gz&quot; &quot;nr.72.tar.gz&quot;
[33] &quot;nr.73.tar.gz&quot; &quot;nr.74.tar.gz&quot; &quot;nr.07.tar.gz&quot; &quot;nr.77.tar.gz&quot;
[37] &quot;nr.78.tar.gz&quot; &quot;nr.08.tar.gz&quot; &quot;nr.09.tar.gz&quot; &quot;nr.10.tar.gz&quot;
[41] &quot;nr.79.tar.gz&quot; &quot;nr.80.tar.gz&quot; &quot;nr.81.tar.gz&quot; &quot;nr.82.tar.gz&quot;
[45] &quot;nr.84.tar.gz&quot; &quot;nr.11.tar.gz&quot; &quot;nr.12.tar.gz&quot; &quot;nr.85.tar.gz&quot;
[49] &quot;nr.13.tar.gz&quot; &quot;nr.86.tar.gz&quot; &quot;nr.87.tar.gz&quot; &quot;nr.14.tar.gz&quot;
[53] &quot;nr.28.tar.gz&quot; &quot;nr.29.tar.gz&quot; &quot;nr.31.tar.gz&quot; &quot;nr.17.tar.gz&quot;
[57] &quot;nr.18.tar.gz&quot; &quot;nr.19.tar.gz&quot; &quot;nr.20.tar.gz&quot; &quot;nr.21.tar.gz&quot;
[61] &quot;nr.22.tar.gz&quot; &quot;nr.23.tar.gz&quot; &quot;nr.32.tar.gz&quot; &quot;nr.24.tar.gz&quot;
[65] &quot;nr.25.tar.gz&quot; &quot;nr.26.tar.gz&quot; &quot;nr.27.tar.gz&quot; &quot;nr.33.tar.gz&quot;
[69] &quot;nr.34.tar.gz&quot; &quot;nr.35.tar.gz&quot; &quot;nr.36.tar.gz&quot; &quot;nr.37.tar.gz&quot;
[73] &quot;nr.38.tar.gz&quot; &quot;nr.39.tar.gz&quot; &quot;nr.40.tar.gz&quot; &quot;nr.41.tar.gz&quot;
[77] &quot;nr.42.tar.gz&quot; &quot;nr.43.tar.gz&quot; &quot;nr.44.tar.gz&quot; &quot;nr.45.tar.gz&quot;
[81] &quot;nr.46.tar.gz&quot; &quot;nr.47.tar.gz&quot; &quot;nr.48.tar.gz&quot; &quot;nr.49.tar.gz&quot;
[85] &quot;nr.53.tar.gz&quot; &quot;nr.50.tar.gz&quot; &quot;nr.51.tar.gz&quot; &quot;nr.52.tar.gz&quot;</code></pre>
<p>The output illustrates that the <code>NCBI nr</code> database has been separated into 41 files.</p>
<p>Further examples are:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all NCBI nt files</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;nt&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;nt.00.tar.gz&quot; &quot;nt.01.tar.gz&quot; &quot;nt.02.tar.gz&quot; &quot;nt.03.tar.gz&quot;
 [5] &quot;nt.04.tar.gz&quot; &quot;nt.05.tar.gz&quot; &quot;nt.06.tar.gz&quot; &quot;nt.07.tar.gz&quot;
 [9] &quot;nt.08.tar.gz&quot; &quot;nt.09.tar.gz&quot; &quot;nt.10.tar.gz&quot; &quot;nt.40.tar.gz&quot;
[13] &quot;nt.11.tar.gz&quot; &quot;nt.41.tar.gz&quot; &quot;nt.42.tar.gz&quot; &quot;nt.43.tar.gz&quot;
[17] &quot;nt.44.tar.gz&quot; &quot;nt.45.tar.gz&quot; &quot;nt.46.tar.gz&quot; &quot;nt.47.tar.gz&quot;
[21] &quot;nt.48.tar.gz&quot; &quot;nt.49.tar.gz&quot; &quot;nt.50.tar.gz&quot; &quot;nt.12.tar.gz&quot;
[25] &quot;nt.51.tar.gz&quot; &quot;nt.52.tar.gz&quot; &quot;nt.13.tar.gz&quot; &quot;nt.14.tar.gz&quot;
[29] &quot;nt.53.tar.gz&quot; &quot;nt.54.tar.gz&quot; &quot;nt.55.tar.gz&quot; &quot;nt.56.tar.gz&quot;
[33] &quot;nt.57.tar.gz&quot; &quot;nt.15.tar.gz&quot; &quot;nt.16.tar.gz&quot; &quot;nt.58.tar.gz&quot;
[37] &quot;nt.26.tar.gz&quot; &quot;nt.27.tar.gz&quot; &quot;nt.17.tar.gz&quot; &quot;nt.18.tar.gz&quot;
[41] &quot;nt.28.tar.gz&quot; &quot;nt.29.tar.gz&quot; &quot;nt.19.tar.gz&quot; &quot;nt.20.tar.gz&quot;
[45] &quot;nt.21.tar.gz&quot; &quot;nt.22.tar.gz&quot; &quot;nt.23.tar.gz&quot; &quot;nt.24.tar.gz&quot;
[49] &quot;nt.25.tar.gz&quot; &quot;nt.30.tar.gz&quot; &quot;nt.31.tar.gz&quot; &quot;nt.32.tar.gz&quot;
[53] &quot;nt.33.tar.gz&quot; &quot;nt.34.tar.gz&quot; &quot;nt.35.tar.gz&quot; &quot;nt.36.tar.gz&quot;
[57] &quot;nt.37.tar.gz&quot; &quot;nt.39.tar.gz&quot; &quot;nt.38.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all NCBI ESTs others</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;est_others&quot;</span>)</code></pre></div>
<pre><code> [1] &quot;est_others.00.tar.gz&quot; &quot;est_others.01.tar.gz&quot;
 [3] &quot;est_others.02.tar.gz&quot; &quot;est_others.03.tar.gz&quot;
 [5] &quot;est_others.04.tar.gz&quot; &quot;est_others.05.tar.gz&quot;
 [7] &quot;est_others.06.tar.gz&quot; &quot;est_others.07.tar.gz&quot;
 [9] &quot;est_others.08.tar.gz&quot; &quot;est_others.09.tar.gz&quot;
[11] &quot;est_others.10.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all NCBI RefSeq (only genomes)</span>
<span class="kw">head</span>(biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;refseq_genomic&quot;</span>), <span class="dv">20</span>)</code></pre></div>
<pre><code>[1] &quot;refseq_genomic.218.tar.gz&quot; &quot;refseq_genomic.219.tar.gz&quot;
 [3] &quot;refseq_genomic.220.tar.gz&quot; &quot;refseq_genomic.00.tar.gz&quot; 
 [5] &quot;refseq_genomic.01.tar.gz&quot;  &quot;refseq_genomic.02.tar.gz&quot; 
 [7] &quot;refseq_genomic.03.tar.gz&quot;  &quot;refseq_genomic.04.tar.gz&quot; 
 [9] &quot;refseq_genomic.05.tar.gz&quot;  &quot;refseq_genomic.06.tar.gz&quot; 
[11] &quot;refseq_genomic.07.tar.gz&quot;  &quot;refseq_genomic.08.tar.gz&quot; 
[13] &quot;refseq_genomic.09.tar.gz&quot;  &quot;refseq_genomic.10.tar.gz&quot; 
[15] &quot;refseq_genomic.11.tar.gz&quot;  &quot;refseq_genomic.12.tar.gz&quot; 
[17] &quot;refseq_genomic.13.tar.gz&quot;  &quot;refseq_genomic.14.tar.gz&quot; 
[19] &quot;refseq_genomic.15.tar.gz&quot;  &quot;refseq_genomic.16.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all NCBI RefSeq (only proteomes)</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;refseq_protein&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;refseq_protein.25.tar.gz&quot; &quot;refseq_protein.00.tar.gz&quot;
 [3] &quot;refseq_protein.01.tar.gz&quot; &quot;refseq_protein.02.tar.gz&quot;
 [5] &quot;refseq_protein.03.tar.gz&quot; &quot;refseq_protein.26.tar.gz&quot;
 [7] &quot;refseq_protein.27.tar.gz&quot; &quot;refseq_protein.28.tar.gz&quot;
 [9] &quot;refseq_protein.29.tar.gz&quot; &quot;refseq_protein.04.tar.gz&quot;
[11] &quot;refseq_protein.05.tar.gz&quot; &quot;refseq_protein.06.tar.gz&quot;
[13] &quot;refseq_protein.07.tar.gz&quot; &quot;refseq_protein.08.tar.gz&quot;
[15] &quot;refseq_protein.09.tar.gz&quot; &quot;refseq_protein.14.tar.gz&quot;
[17] &quot;refseq_protein.15.tar.gz&quot; &quot;refseq_protein.16.tar.gz&quot;
[19] &quot;refseq_protein.41.tar.gz&quot; &quot;refseq_protein.10.tar.gz&quot;
[21] &quot;refseq_protein.11.tar.gz&quot; &quot;refseq_protein.17.tar.gz&quot;
[23] &quot;refseq_protein.12.tar.gz&quot; &quot;refseq_protein.13.tar.gz&quot;
[25] &quot;refseq_protein.18.tar.gz&quot; &quot;refseq_protein.19.tar.gz&quot;
[27] &quot;refseq_protein.20.tar.gz&quot; &quot;refseq_protein.21.tar.gz&quot;
[29] &quot;refseq_protein.22.tar.gz&quot; &quot;refseq_protein.23.tar.gz&quot;
[31] &quot;refseq_protein.24.tar.gz&quot; &quot;refseq_protein.30.tar.gz&quot;
[33] &quot;refseq_protein.31.tar.gz&quot; &quot;refseq_protein.32.tar.gz&quot;
[35] &quot;refseq_protein.33.tar.gz&quot; &quot;refseq_protein.34.tar.gz&quot;
[37] &quot;refseq_protein.35.tar.gz&quot; &quot;refseq_protein.36.tar.gz&quot;
[39] &quot;refseq_protein.37.tar.gz&quot; &quot;refseq_protein.38.tar.gz&quot;
[41] &quot;refseq_protein.39.tar.gz&quot; &quot;refseq_protein.40.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show all NCBI RefSeq (only RNA)</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;refseq_rna&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;refseq_rna.00.tar.gz&quot; &quot;refseq_rna.01.tar.gz&quot;
 [3] &quot;refseq_rna.10.tar.gz&quot; &quot;refseq_rna.02.tar.gz&quot;
 [5] &quot;refseq_rna.05.tar.gz&quot; &quot;refseq_rna.03.tar.gz&quot;
 [7] &quot;refseq_rna.06.tar.gz&quot; &quot;refseq_rna.04.tar.gz&quot;
 [9] &quot;refseq_rna.07.tar.gz&quot; &quot;refseq_rna.08.tar.gz&quot;
[11] &quot;refseq_rna.09.tar.gz&quot; &quot;refseq_rna.11.tar.gz&quot;
[13] &quot;refseq_rna.12.tar.gz&quot; &quot;refseq_rna.13.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show NCBI swissprot</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;swissprot&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;swissprot.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show NCBI PDB</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;pdb&quot;</span>)</code></pre></div>
<pre><code> [1] &quot;pdbnt.00.tar.gz&quot; &quot;pdbnt.40.tar.gz&quot; &quot;pdbnt.41.tar.gz&quot;
 [4] &quot;pdbnt.42.tar.gz&quot; &quot;pdbnt.43.tar.gz&quot; &quot;pdbnt.44.tar.gz&quot;
 [7] &quot;pdbnt.45.tar.gz&quot; &quot;pdbnt.46.tar.gz&quot; &quot;pdbnt.47.tar.gz&quot;
[10] &quot;pdbnt.48.tar.gz&quot; &quot;pdbnt.49.tar.gz&quot; &quot;pdbnt.50.tar.gz&quot;
[13] &quot;pdbnt.51.tar.gz&quot; &quot;pdbnt.52.tar.gz&quot; &quot;pdbnt.53.tar.gz&quot;
[16] &quot;pdbnt.54.tar.gz&quot; &quot;pdbnt.55.tar.gz&quot; &quot;pdbnt.56.tar.gz&quot;
[19] &quot;pdbnt.57.tar.gz&quot; &quot;pdbnt.58.tar.gz&quot; &quot;pdbnt.26.tar.gz&quot;
[22] &quot;pdbnt.27.tar.gz&quot; &quot;pdbnt.01.tar.gz&quot; &quot;pdbnt.02.tar.gz&quot;
[25] &quot;pdbnt.03.tar.gz&quot; &quot;pdbnt.04.tar.gz&quot; &quot;pdbnt.05.tar.gz&quot;
[28] &quot;pdbnt.06.tar.gz&quot; &quot;pdbnt.07.tar.gz&quot; &quot;pdbnt.08.tar.gz&quot;
[31] &quot;pdbnt.09.tar.gz&quot; &quot;pdbnt.10.tar.gz&quot; &quot;pdbnt.11.tar.gz&quot;
[34] &quot;pdbnt.12.tar.gz&quot; &quot;pdbnt.13.tar.gz&quot; &quot;pdbnt.14.tar.gz&quot;
[37] &quot;pdbnt.15.tar.gz&quot; &quot;pdbnt.16.tar.gz&quot; &quot;pdbnt.17.tar.gz&quot;
[40] &quot;pdbnt.18.tar.gz&quot; &quot;pdbnt.28.tar.gz&quot; &quot;pdbnt.19.tar.gz&quot;
[43] &quot;pdbnt.20.tar.gz&quot; &quot;pdbnt.21.tar.gz&quot; &quot;pdbnt.22.tar.gz&quot;
[46] &quot;pdbnt.23.tar.gz&quot; &quot;pdbnt.24.tar.gz&quot; &quot;pdbnt.29.tar.gz&quot;
[49] &quot;pdbnt.25.tar.gz&quot; &quot;pdbaa.tar.gz&quot;    &quot;pdbnt.30.tar.gz&quot;
[52] &quot;pdbnt.31.tar.gz&quot; &quot;pdbnt.32.tar.gz&quot; &quot;pdbnt.33.tar.gz&quot;
[55] &quot;pdbnt.34.tar.gz&quot; &quot;pdbnt.35.tar.gz&quot; &quot;pdbnt.36.tar.gz&quot;
[58] &quot;pdbnt.37.tar.gz&quot; &quot;pdbnt.39.tar.gz&quot; &quot;pdbnt.38.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show NCBI Human database</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;human&quot;</span>)</code></pre></div>
<pre><code>[1] &quot;human_genomic.00.tar.gz&quot; &quot;human_genomic.01.tar.gz&quot;
 [3] &quot;human_genomic.02.tar.gz&quot; &quot;human_genomic.03.tar.gz&quot;
 [5] &quot;human_genomic.04.tar.gz&quot; &quot;human_genomic.05.tar.gz&quot;
 [7] &quot;human_genomic.06.tar.gz&quot; &quot;human_genomic.07.tar.gz&quot;
 [9] &quot;human_genomic.08.tar.gz&quot; &quot;human_genomic.10.tar.gz&quot;
[11] &quot;human_genomic.11.tar.gz&quot; &quot;human_genomic.12.tar.gz&quot;
[13] &quot;human_genomic.13.tar.gz&quot; &quot;human_genomic.14.tar.gz&quot;
[15] &quot;human_genomic.15.tar.gz&quot; &quot;human_genomic.16.tar.gz&quot;
[17] &quot;human_genomic.18.tar.gz&quot; &quot;human_genomic.19.tar.gz&quot;
[19] &quot;human_genomic.20.tar.gz&quot; &quot;human_genomic.21.tar.gz&quot;
[21] &quot;human_genomic.22.tar.gz&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show NCBI EST databases</span>
biomartr<span class="op">::</span><span class="kw">listNCBIDatabases</span>(<span class="dt">db =</span> <span class="st">&quot;est&quot;</span>)</code></pre></div>
<pre><code> [1] &quot;est.tar.gz&quot;           &quot;est_human.00.tar.gz&quot; 
 [3] &quot;est_human.01.tar.gz&quot;  &quot;est_mouse.tar.gz&quot;    
 [5] &quot;est_others.00.tar.gz&quot; &quot;est_others.01.tar.gz&quot;
 [7] &quot;est_others.02.tar.gz&quot; &quot;est_others.03.tar.gz&quot;
 [9] &quot;est_others.04.tar.gz&quot; &quot;est_others.05.tar.gz&quot;
[11] &quot;est_others.06.tar.gz&quot; &quot;est_others.07.tar.gz&quot;
[13] &quot;est_others.08.tar.gz&quot; &quot;est_others.09.tar.gz&quot;
[15] &quot;est_others.10.tar.gz&quot;</code></pre>
<p><strong>Please not that all lookup and retrieval function will only work properly when a sufficient internet connection is provided.</strong></p>
<p>In a next step users can use the <code>listNCBIDatabases()</code> and <code>download.database.all()</code> functions to retrieve all files corresponding to a specific NCBI database.</p>
</div>
<div id="download-ncbi-databases" class="section level2">
<h2>Download NCBI databases</h2>
<p>Using the same search strategy by specifying the database name as described above, users can now download these databases using the <code>download.database.all()</code> function.</p>
<p>For downloading only single files users can type:</p>
<div id="example-ncbi-nr" class="section level3">
<h3>Example NCBI nr</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI nr database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;nr&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;nr&quot;</span>)</code></pre></div>
<p>This command will download the pre-formatted (by makeblastdb formatted) database version is retrieved.</p>
<p>Using this command, all <code>NCBI nr</code> files are loaded into the <code>nr</code> folder (<code>path = &quot;nr&quot;</code>). For each data package, <code>biomartr</code> checks the <code>md5checksum</code> of the downloaded file and the file stored online to make sure that internet connection losses didn’t currupt the file. In case you see a warning message notifying you about not-matching <code>md5checksum</code> values, please re-download the corresponding data package by re-running the <code>download.database.all()</code> command. From my own experience this can happen when server connections or internet connections are not very stable during the download process of large data chunks.</p>
</div>
<div id="example-ncbi-nt" class="section level3">
<h3>Example NCBI nt</h3>
<p>The same approach can be applied to all other databases mentioned above, e.g.:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI nt database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;nt&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;nt&quot;</span>)</code></pre></div>
</div>
<div id="example-ncbi-refseq" class="section level3">
<h3>Example NCBI RefSeq</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI refseq (protein) database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;refseq_protein&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;refseq_protein&quot;</span>)</code></pre></div>
</div>
<div id="example-pdb" class="section level3">
<h3>Example PDB</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI PDB database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;pdb&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;pdb&quot;</span>)</code></pre></div>
</div>
<div id="example-ncbi-taxonomy" class="section level3">
<h3>Example NCBI Taxonomy</h3>
<p>Download NCBI Taxonomy via:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI taxonomy database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;taxdb&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;taxdb&quot;</span>)</code></pre></div>
<pre><code>Starting download of the files: taxdb.tar.gz, taxdb.btd, taxdb.bti ...
This download process may take a while due to the large size of the individual data chunks ...
Starting download process of file: taxdb.tar.gz ...
Checking md5 hash of file: taxdb.tar.gz ...
The md5 hash of file 'taxdb.tar.gz' matches!
File 'taxdb/taxdb.tar.gz has successfully been retrieved.
Starting download process of file: taxdb.btd ...
Checking md5 hash of file: taxdb.btd ...
The md5 hash of file 'taxdb.btd' matches!
File 'taxdb/taxdb.btd has successfully been retrieved.
Starting download process of file: taxdb.bti ...
Checking md5 hash of file: taxdb.bti ...
The md5 hash of file 'taxdb.bti' matches!
File 'taxdb/taxdb.bti has successfully been retrieved.
Download process is finished and files are stored in 'taxdb'.</code></pre>
</div>
<div id="example-ncbi-swissprot" class="section level3">
<h3>Example NCBI Swissprot</h3>
<p>Download NCBI Swissprot via:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI swissprot database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;swissprot&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;swissprot&quot;</span>)</code></pre></div>
</div>
<div id="example-ncbi-cdd-delta" class="section level3">
<h3>Example NCBI CDD Delta</h3>
<p>Download NCBI CDD Delta via:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># download the entire NCBI CDD Delta database</span>
biomartr<span class="op">::</span><span class="kw">download.database.all</span>(<span class="dt">db =</span> <span class="st">&quot;cdd_delta&quot;</span>, <span class="dt">path =</span> <span class="st">&quot;cdd_delta&quot;</span>)</code></pre></div>
<p>For each data package, <code>biomartr</code> checks the <code>md5checksum</code> of the downloaded file and the file stored online to make sure that internet connection losses didn’t currupt the file. In case you see a warning message notifying you about not-matching <code>md5checksum</code> values, please re-download the corresponding data package. From my own experience this can happen when server connections or internet connections are not very stable during the download process of large data chunks.</p>
<p><strong>Please notice that most of these databases are very large, so users should take of of providing a stable internet connection throughout the download process.</strong></p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
